### PR TITLE
RNRBC - 391 Date error messages improvement

### PR DESCRIPTION
**Pull request by tyagikumarvineet**

### CODE REVIEW

File: app/uk/gov/hmrc/residencenilratebandcalculator/utils/FormHelpers.scala [hunk 1]
```diff
@@ -30,7 +32,7 @@ object FormHelpers {
     either.merge
   }
 
-  def getDatePart(form: Option[Form[Date]], datePart: DatePart) = {
+/*  def getDatePart(form: Option[Form[Date]], datePart: DatePart) = {
```
Comment (grantrobson) [inline app/uk/gov/hmrc/residencenilratebandcalculator/utils/FormHelpers.scala]: Remove comment

File: app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_fields.scala.html [hunk 2]
```diff
@@ -0,0 +1,96 @@
+@*
+* Copyright 2016 HM Revenue & Customs
+*
```
Comment (grantrobson) [inline app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_fields.scala.html]: Indentation not right

File: app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_input.scala.html [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*@
+
+@(field: play.api.data.Field, args: (Symbol,Any)*)(implicit messages: Messages)
+
```
Comment (grantrobson) [inline app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_input.scala.html]: Indentation not right

File: conf/messages.en [hunk 4]
```diff
@@ -57,6 +58,20 @@ date_of_death.error = Give a correct date of death
 date_of_death.error.day_invalid = Give a correct day of death using numbers 1 to 31
 date_of_death.error.month_invalid = Give a correct month of death using numbers 1 to 12
 date_of_death.error.year_invalid = Give a correct year of death
+date_Of_death.error.date_not_complete = Give a full date of death
```
Comment (grantrobson) [inline conf/messages.en]: Need to get the Welsh translations for new content - email gareth.jones4@hmrc.gsi.gov.uk

File: app/uk/gov/hmrc/residencenilratebandcalculator/forms/mappings/DateMapping.scala [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+            } else if (!isDayValidPredicate(day)) {
+              Invalid(errorInvalidDayKey)
+            } else {
+              Valid
+            }
+        }
+}
+    )
+
+  private def checkDateElementsMakeValidNonFutureDate(dateAsTuple: (String, String, String),
```
Comment (jennygrahamjones) [inline app/uk/gov/hmrc/residencenilratebandcalculator/forms/mappings/DateMapping.scala]: It doesn't look like this function is ever used. I don't think it's relevant for RNRB.

File: app/uk/gov/hmrc/residencenilratebandcalculator/forms/mappings/DateMapping.scala [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+  /**
+    * errorBlankFieldKey - if the date, or any portion of it, is blank
+    * errorInvalidCharsKey - if any invalid chars in date
+    * errorInvalidDayKey - if the day portion of the date is numeric but invalid, e.g. 33
+    * errorInvalidDayForMonthKey - if the day portion of the date is numeric but invalid for the month, e.g. 30 for month of 2
+    * errorInvalidMonthKey - if the month portion of the date is numeric but invalid, e.g. 13
+    * errorInvalidYearKey - if the year potion of the date is numeric but of less than 4 digits
+    * errorInvalidAllKey - if all portions of the date are numeric but invalid as described above
+    * errorDateInFutureKey - if the date is in the future
```
Comment (jennygrahamjones) [inline app/uk/gov/hmrc/residencenilratebandcalculator/forms/mappings/DateMapping.scala]: line 134 refers to a non-existent parameter

File: app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_input.scala.html [hunk 7]
```diff
@@ -0,0 +1,77 @@
+@*
+* Copyright 2016 HM Revenue & Customs
```
Comment (jennygrahamjones) [inline app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_input.scala.html]: Is it not possible use / adapt the existing input_date view helper? This looks a lot more complicated. If not, the input_date helper is no longer in use so should probably be deleted.

Also, is there a reason for splitting the date inputs across date_input and date_fields?

File: app/uk/gov/hmrc/residencenilratebandcalculator/controllers/DateOfDeathController.scala [hunk 8]
```diff
@@ -41,26 +41,28 @@ class DateOfDeathController @Inject()(val appConfig: FrontendAppConfig,
...
+        val cacheMap: CacheMap = optionalCacheMap.getOrElse(CacheMap(hc.sessionId.getOrElse(SessionId("")).value, Map()))
+        val dateOfDeath = cacheMap.getEntry[Date](controllerId)
+
+        Ok(view(dateOfDeath.fold(form)(value => form.fill(value))))
       })
   }
 
-  def onSubmit(implicit wts: Writes[Date]) = Action.async { implicit request =>
-    val boundForm = form().bindFromRequest()
+  def onSubmit(implicit rds: Writes[Date]) = Action.async { implicit request =>
```
Comment (jennygrahamjones) [inline app/uk/gov/hmrc/residencenilratebandcalculator/controllers/DateOfDeathController.scala]: why is this 'rds' now? the type is still Writes[Date], not Reads[Date]

File: app/uk/gov/hmrc/residencenilratebandcalculator/forms/DateForm.scala [hunk 9]
```diff
@@ -16,44 +16,22 @@
...
-import play.api.data.format.Formatter
+import uk.gov.hmrc.residencenilratebandcalculator.forms.mappings.DateMapping
 import uk.gov.hmrc.residencenilratebandcalculator.models.Date
-import uk.gov.hmrc.residencenilratebandcalculator.forms.FormValidators._
 
 object DateForm {
 
-  val upperBoundDays = 31
-  val upperBoundMonths = 12
-  val upperBoundYears = 2050
```
Comment (jennygrahamjones) [inline app/uk/gov/hmrc/residencenilratebandcalculator/forms/DateForm.scala]: this upper bound doesn't seem to have been included in the revised code

File: app/uk/gov/hmrc/residencenilratebandcalculator/utils/FormHelpers.scala (hunk 1 above)
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/residencenilratebandcalculator/utils/FormHelpers.scala]: Removed comments

File: app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_fields.scala.html (hunk 2 above)
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_fields.scala.html]: Done

File: app/uk/gov/hmrc/residencenilratebandcalculator/forms/mappings/DateMapping.scala (hunk 5 above)
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/residencenilratebandcalculator/forms/mappings/DateMapping.scala]: Deleted

File: app/uk/gov/hmrc/residencenilratebandcalculator/forms/mappings/DateMapping.scala (hunk 6 above)
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/residencenilratebandcalculator/forms/mappings/DateMapping.scala]: Done

File: app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_input.scala.html (hunk 7 above)
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_input.scala.html]: Yes, it has been done like IHT that is why it has been used.

Comment (jennygrahamjones) [inline app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_input.scala.html]: is there a reason why we couldn't adapt the existing code?

File: app/uk/gov/hmrc/residencenilratebandcalculator/controllers/DateOfDeathController.scala (hunk 8 above)
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/residencenilratebandcalculator/controllers/DateOfDeathController.scala]: Done

File: conf/messages.en (hunk 4 above)
Comment (tyagikumarvineet) [inline conf/messages.en]: Have sent mail to Gareth to get the translations

File: app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_input.scala.html (hunk 7 above)
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/residencenilratebandcalculator/views/rnrbHelpers/date_input.scala.html]: Code refactored

File: app/uk/gov/hmrc/residencenilratebandcalculator/forms/DateForm.scala (hunk 9 above)
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/residencenilratebandcalculator/forms/DateForm.scala]: Has been taken care in DateMappings and outbound limit is 31/12/9999

**Comment by jennygrahamjones** [comment]
4 unit tests on DateFormSpec are failing

File: conf/messages.en (hunk 4 above)
Comment (tyagikumarvineet) [inline conf/messages.en]: Welsh translation is done.

**Pull request merged**